src/app/shared/forbidden-username.directive.ts

```diff
@@ -0,0 +1,28 @@
+import { Directive, Input } from '@angular/core';
+import { AbstractControl, NG_VALIDATORS, Validator, ValidatorFn } from '@angular/forms';
+
+@Directive({
+    selector: '[appForbiddenUsername]',
+    providers: [{ provide: NG_VALIDATORS, useExisting: ForbiddenUsernameDirective, multi: true }],
+})
+export class ForbiddenUsernameDirective implements Validator {
+    @Input('appForbiddenUsername') forbiddenChar: string;
+
+    validate(control: AbstractControl): { [key: string]: any } | null {
+        // if forbiddenChar is present, call Validator function forbiddenNameValidator with regexp parameter (case insensitive) and its inner function with control, which is input value)
+        return this.forbiddenChar
+            ? forbiddenNameValidator(new RegExp(this.forbiddenChar, 'i'))(control)
+            : null;
+    }
+}
+
+/* function forbiddenNameValidator takes regular expression (later made from appForbiddenUsername) as parameter and returns a validator function, 
```

Comment:
The first two line of the comment is completely redundant. All of this description is done by the typing and naming.
```
// function takes regex as parameter and returns validator function
export function forbiddenNameValidator(regex: RegExp): ValidatorFn
```

Furthermore, the rest of the comment also describes the inner function's work, which can be derived from how it works.

I'm going to leave this comment here and approve this, and we will see in January if you will be able to derive what the function does from this comment or if it will be rather misleading. There are a few, but I'm approving this to show it in January.